File: backend/app/schemas/trip.py
from datetime import datetime
from pydantic import BaseModel, Field


class TripStart(BaseModel):
    boat_id: int | None = None
    start_latitude: float | None = Field(default=None, ge=-90, le=90)
    start_longitude: float | None = Field(default=None, ge=-180, le=180)
    destination: str | None = None
    estimated_return_at: datetime | None = None
    notes: str | None = None


class TripEnd(BaseModel):
    notes: str | None = None


class TripTransition(BaseModel):
    status: str = Field(..., pattern="^(planned|active|returning|completed|cancelled|emergency)$")
    notes: str | None = None


class TripOut(BaseModel):
    id: int
    user_id: int
    boat_id: int | None
    status: str
    start_time: datetime
    end_time: datetime | None
    estimated_return_at: datetime | None
    start_latitude: float | None
    start_longitude: float | None
    destination: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: mobile/lib/models/trip.dart
/// Mirrors backend TripOut (app/schemas/trip.py) and the state machine in
/// app/services/trip_service.py. Status is one of: planned, active,
/// returning, completed, cancelled, emergency.
class Trip {
  final int id;
  final String status;
  final DateTime startTime;
  final DateTime? endTime;
  final DateTime? estimatedReturnAt;
  final String? destination;
  final String? notes;

  Trip({
    required this.id,
    required this.status,
    required this.startTime,
    this.endTime,
    this.estimatedReturnAt,
    this.destination,
    this.notes,
  });

  bool get isInProgress => status == 'active' || status == 'returning' || status == 'emergency';

  factory Trip.fromJson(Map<String, dynamic> json) => Trip(
        id: json['id'] as int,
        status: json['status'] as String,
        startTime: DateTime.parse(json['start_time'] as String),
        endTime: json['end_time'] == null ? null : DateTime.parse(json['end_time'] as String),
        estimatedReturnAt:
            json['estimated_return_at'] == null ? null : DateTime.parse(json['estimated_return_at'] as String),
        destination: json['destination'] as String?,
        notes: json['notes'] as String?,
      );
}
